source/common/http/conn_manager_impl.cc

```diff
@@ -1207,23 +1211,31 @@ void ConnectionManagerImpl::ActiveStream::encodeHeaders(ActiveStreamEncoderFilte
 }
 
 void ConnectionManagerImpl::ActiveStream::encodeMetadata(ActiveStreamEncoderFilter* filter,
-                                                         MetadataMapPtr&& metadata_map) {
+                                                         MetadataMapPtr&& metadata_map_ptr) {
   resetIdleTimer();
 
-  std::list<ActiveStreamEncoderFilterPtr>::iterator entry = commonEncodePrefix(filter, false);
+  if (current_encoder_filter_ != nullptr) {
+    // current_encoder_filter_ != nullptr means new metadata is added locally by
+    // current_encoder_filter_. Pass the new metadata through all downstream filters of
+    // current_encoder_filter_.
+    filter = current_encoder_filter_;
+  }
+  std::list<ActiveStreamEncoderFilterPtr>::iterator entry =
+      filter == nullptr ? encoder_filters_.begin() : std::next(filter->entry());
   for (; entry != encoder_filters_.end(); entry++) {
-    // TODO(soya3129): Add filters here.
+    FilterMetadataStatus status = (*entry)->handle_->encodeMetadata(*metadata_map_ptr);
```

Comment:
current_encoder_filter_ not updated here.  what if one of these filters calls encodeMetadata. Should that follow-up encode only go downstream, or should we fail because this is not supported?

Again I'm entirely fine just always passing metadata through all the filters for now and fixing later.

Reply:
Good catch!

Fixed by updating current_encoder_filter_, and let the newly added metadata to go through downstream filters only. Also added a test case to verify the change.

I think if we would like StreamEncoderFilter::encodeMetadata() to be able to add new metadata through StreamDecoderFilterCallback::encodeMetadata(), we will have to pass metadata through downstream filters only. Otherwise, we may end up with an infinite loop (StreamEncoderFilter::encodeMetadata() -> StreamDecoderFilterCallback::encodeMetadata() -> StreamEncoderFilter::encodeMetadata()). If we eventually decide to let new metadata pass all filters, StreamEncoderFilter::encodeMetadata(MetadataMap& metadata_map) has to add new metadata to metadata_map instead of calling StreamDecoderFilterCallback::encodeMetadata().

Reply:
Is the question about why we need to save current encoder filter in current_encoder_filter_ for metadata, but not for other data type?

That is because we call StreamDecoderFilterCallbacks::encodeMetadata() in FooFilter::encodeXXX(), and there is no way to pass the current filter's pointer through the callback. The call stack to parse metadata is: FooFilter::EncodeXXX() calls StreamDecoderFilterCallbacks::encodeMetadata() which calls ConnectionManagerImpl::ActiveStreamDecoderFilter::encodeMetadata(), which calls: ActiveStream::encodeMetadata().

I think other data types don't need to save the current encoder pointer because they either want the data to go through all the encoder filters (For example, router filter calls callbacks_->encodeXXX() after decoding the data), or they call commonContinue() or similar functions of the current encoder filter which only pass the data through downstream filters. encodeMetadata() only returns continue, so we may be able to avoid the complexity of commonContinue() logic.

Storing current_encoder_filter_ is only needed for response metadata. Request metadata doesn't need this logic.

Reply:
I see, OK. IMO this is a larger problem that needs thinking about and my preference is to solve this a different way. I.e., the way I would envision this working is that if an encoder filter wants data to be encoded only to "earlier" filters, it should call through `StreamEncoderFilterCallbacks`. I think the same thing applies also to the other encoding types. Given that I think this is a larger problem than just for metadata, WDYT about just unwinding this part of the change, and going with @alyssawilk's suggestion to just have it go through all filters for now and then we can think about fixing this in a discrete change? I think this is a meaty issue that we need to think through and I would rather do it independently.

Reply:
I see. Sure! Will unwind this part. Thanks!

I would like to discuss and learn more about the future changes around this area, :)

Reply:
nope!